include/options/option/abstract-option.php

```diff
@@ -0,0 +1,316 @@
+<?php
+/**
+ * @package Polylang
+ */
+
+namespace WP_Syntex\Polylang\Options\Option;
+
+use WP_Error;
+use WP_Syntex\Polylang\Options\Options;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Class defining a single option.
+ *
+ * @since 3.7
+ *
+ * @phpstan-type SchemaType 'string'|'null'|'number'|'integer'|'boolean'|'array'|'object'
+ * @phpstan-type Schema array{
+ *     '$schema': non-falsy-string,
+ *     title: non-falsy-string,
+ *     description: string,
+ *     type: SchemaType,
+ *     context: array<non-falsy-string>
+ * }&array<non-falsy-string, mixed>
+ */
+abstract class Abstract_Option {
+	/**
+	 * Option key.
+	 *
+	 * @var string
+	 * @phpstan-var non-falsy-string
+	 */
+	private $key;
+
+	/**
+	 * Option value.
+	 *
+	 * @var mixed
+	 */
+	private $value;
+
+	/**
+	 * Option default value.
+	 *
+	 * @var mixed
+	 */
+	private $default;
+
+	/**
+	 * Cached option JSON schema.
+	 *
+	 * @var array|null
+	 */
+	private $schema;
+
+	/**
+	 * Option description.
+	 *
+	 * @var string
+	 */
+	private $description;
+
+	/**
+	 * Validation and sanitization errors.
+	 *
+	 * @var WP_Error
+	 */
+	protected $errors;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 3.7
+	 *
+	 * @param string $key         Option key.
+	 * @param mixed  $value       Option value.
+	 * @param mixed  $default     Option default value.
+	 * @param string $description Option description, used in JSON schema.
+	 *
+	 * @phpstan-param non-falsy-string $key
+	 */
+	public function __construct( string $key, $value, $default, string $description ) {
+		$this->errors      = new WP_Error();
+		$this->key         = $key;
+		$this->default     = $default;
+		$this->description = $description;
+
+		$value = rest_sanitize_value_from_schema( $value, $this->get_schema(), $this->key() );
+
+		if ( ! is_wp_error( $value ) ) {
+			$this->value = $value;
+		} else {
+			$this->value = $default;
+		}
+	}
+
+	/**
+	 * Returns option key.
+	 *
+	 * @since 3.7
+	 *
+	 * @return string
+	 *
+	 * @phpstan-return non-falsy-string
+	 */
+	public function key(): string {
+		return $this->key;
+	}
+
+	/**
+	 * Sets option's value if valid, does nothing otherwise.
+	 *
+	 * @since 3.7
+	 *
+	 * @param mixed   $value   Value to set.
+	 * @param Options $options All options.
+	 * @return bool True if the value has been assigned. False in case of errors.
+	 */
+	public function set( $value, Options $options ): bool {
+		$this->errors = new WP_Error(); // Reset errors.
+
+		if ( ! $this->validate( $value, $options ) ) {
+			// Blocking validation error.
+			return false;
+		}
+
+		$value = $this->sanitize( $value, $options );
+
+		if ( $this->has_blocking_errors() ) {
+			// Blocking sanitization error.
+			return false;
+		}
```

Comment:
I suggest that you return the `WP_Error` when it's blocking:
```suggestion
		if ( is_wp_error( $value ) ) {
			// Blocking sanitization error.
			return false;
		}
```